From fddfa101f128ffb6a28f39680ad3779ac7cd6b47 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Thu, 23 Oct 2025 12:43:39 +0100
Subject: [PATCH 2/2] rand: for cryptographically secure operations use std
 rand

Many go toolchains (boring, MS systemcrypto, FIPS140 mode) require to
use approved random number generator (crypto/rand.Reader) and fail to
generate keys and signatures if any other random number generator is
supplied (including wrappers).

Thus ensure that keys, certificates, signatures are generated using
standard rand.Reader always.
---
 lib/signature/signature.go | 3 +--
 lib/tlsutil/tlsutil.go     | 5 +++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/lib/signature/signature.go b/lib/signature/signature.go
index de5c025ee..95ac0c598 100644
--- a/lib/signature/signature.go
+++ b/lib/signature/signature.go
@@ -11,6 +11,7 @@ package signature
 import (
 	"crypto/ecdsa"
 	"crypto/elliptic"
+	"crypto/rand"
 	"crypto/sha256"
 	"crypto/x509"
 	"encoding/asn1"
@@ -19,8 +20,6 @@ import (
 	"fmt"
 	"io"
 	"math/big"
-
-	"github.com/syncthing/syncthing/lib/rand"
 )
 
 // GenerateKeys returns a new key pair, with the private and public key
diff --git a/lib/tlsutil/tlsutil.go b/lib/tlsutil/tlsutil.go
index d9a54c452..61f5bfa10 100644
--- a/lib/tlsutil/tlsutil.go
+++ b/lib/tlsutil/tlsutil.go
@@ -11,6 +11,7 @@ import (
 	"crypto/ed25519"
 	"crypto/elliptic"
 	"crypto/fips140"
+	"crypto/rand"
 	"crypto/rsa"
 	"crypto/tls"
 	"crypto/x509"
@@ -23,7 +24,7 @@ import (
 	"os"
 	"time"
 
-	"github.com/syncthing/syncthing/lib/rand"
+	srand "github.com/syncthing/syncthing/lib/rand"
 )
 
 var (
@@ -127,7 +128,7 @@ func generateCertificate(commonName string, lifetimeDays int, compatible bool) (
 	// you add or change attributes in here, especially DNSNames or
 	// IPAddresses.
 	template := x509.Certificate{
-		SerialNumber: new(big.Int).SetUint64(rand.Uint64()),
+		SerialNumber: new(big.Int).SetUint64(srand.Uint64()),
 		Subject: pkix.Name{
 			CommonName:         commonName,
 			Organization:       []string{"Syncthing"},
-- 
2.51.0

